flight/Libraries/insgps14state.c

```diff
@@ -0,0 +1,905 @@
+		Rbe_a[1][2] = k1*sqrtf(-powf(q0*q2*2.0f-q1*q3*2.0f,2.0f)+1.0f)*(q0*q1*2.0f+q2*q3*2.0f);
+		Rbe_a[2][0] = k1*(q0*q2*2.0f-q1*q3*2.0f)*(q0*q0-q1*q1-q2*q2+q3*q3);
+		Rbe_a[2][1] = -k1*(q0*q1*2.0f+q2*q3*2.0f);
+		Rbe_a[2][2] = k1*k2*(q0*q0-q1*q1-q2*q2+q3*q3);
+
+		Z[6] = Rbe_a[0][0]*mag_data[0] + Rbe_a[1][0]*mag_data[1] + Rbe_a[2][0]*mag_data[2] ;
+		Z[7] = Rbe_a[0][1]*mag_data[0] + Rbe_a[1][1]*mag_data[1] + Rbe_a[2][1]*mag_data[2] ;
+		Z[8] = Rbe_a[0][2]*mag_data[0] + Rbe_a[1][2]*mag_data[1] + Rbe_a[2][2]*mag_data[2] ;
+	}
+
+	// barometric altimeter in meters and in local NED frame
+	Z[9] = BaroAlt;
+
+	// EKF correction step
+	LinearizeH(X, Be, H);
+	MeasurementEq(X, Be, Y);
+	SerialUpdate(H, R, Z, Y, P, X, SensorsUsed);
+	qmag = sqrtf(X[6] * X[6] + X[7] * X[7] + X[8] * X[8] + X[9] * X[9]);
+	X[6] /= qmag;
+	X[7] /= qmag;
+	X[8] /= qmag;
+	X[9] /= qmag;
+}
+
+//  *************  CovariancePrediction *************
+//  Does the prediction step of the Kalman filter for the covariance matrix
+//  Output, Pnew, overwrites P, the input covariance
+//  Pnew = (I+F*T)*P*(I+F*T)' + T^2*G*Q*G'
+//  Q is the discrete time covariance of process noise
+//  Q is vector of the diagonal for a square matrix with
+//    dimensions equal to the number of disturbance noise variables
+//  The General Method is very inefficient,not taking advantage of the sparse F and G
+//  The first Method is very specific to this implementation
+//  ************************************************
+
+#ifdef COVARIANCE_PREDICTION_GENERAL
+
+void CovariancePrediction(float F[NUMX][NUMX], float G[NUMX][NUMW],
+			  float Q[NUMW], float dT, float P[NUMX][NUMX])
+{
+	float Dummy[NUMX][NUMX], dTsq;
+	uint8_t i, j, k;
+
+	//  Pnew = (I+F*T)*P*(I+F*T)' + T^2*G*Q*G' = T^2[(P/T + F*P)*(I/T + F') + G*Q*G')]
+
+	dTsq = dT * dT;
+
+	for (i = 0; i < NUMX; i++)	// Calculate Dummy = (P/T +F*P)
+		for (j = 0; j < NUMX; j++) {
+			Dummy[i][j] = P[i][j] / dT;
+			for (k = 0; k < NUMX; k++)
+				Dummy[i][j] += F[i][k] * P[k][j];
+		}
+	for (i = 0; i < NUMX; i++)	// Calculate Pnew = Dummy/T + Dummy*F' + G*Qw*G'
+		for (j = i; j < NUMX; j++) {	// Use symmetry, ie only find upper triangular
+			P[i][j] = Dummy[i][j] / dT;
+			for (k = 0; k < NUMX; k++)
+				P[i][j] += Dummy[i][k] * F[j][k];	// P = Dummy/T + Dummy*F'
+			for (k = 0; k < NUMW; k++)
+				P[i][j] += Q[k] * G[i][k] * G[j][k];	// P = Dummy/T + Dummy*F' + G*Q*G'
+			P[j][i] = P[i][j] = P[i][j] * dTsq;	// Pnew = T^2*P and fill in lower triangular;
+		}
+}
+
+#else
+
+void CovariancePrediction(float F[NUMX][NUMX], float G[NUMX][NUMW],
+			  float Q[NUMW], float dT, float P[NUMX][NUMX])
+{
+	float D[NUMX][NUMX], T, Tsq;
+	uint8_t i, j;
+
+	//  Pnew = (I+F*T)*P*(I+F*T)' + T^2*G*Q*G' = scalar expansion from symbolic manipulator
+
+	T = dT;
+	Tsq = dT * dT;
+
+	for (i = 0; i < NUMX; i++)	// Create a copy of the upper triangular of P
+		for (j = i; j < NUMX; j++)
+			D[i][j] = P[i][j];
+
+	// Brute force calculation of the elements of P
```

Comment:
That's pretty old code from the 13 state INS, just generic matrix multiplication unrolled.